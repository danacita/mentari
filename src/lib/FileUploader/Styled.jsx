import styled, { css } from 'styled-components';
import Color from '../Styles/bases/Color';

const InputWrapper = styled.div`
  position: relative;
  p {
    margin-bottom: 8px;
    &.help-block {
      font-weight: 400;
      font-style: italic;
      margin-bottom: 20px;
    }
  }
  input[type='file'] {
    width: 0.1px;
    height: 0.1px;
    opacity: 0;
    overflow: hidden;
    position: absolute;
    z-index: -1;
    + label {
      font-size: 16px;
      cursor: pointer;
      padding: 8px 16px;
      display: block;
      border-radius: 4px;
      margin: 8px 0 16px 0;
      overflow: hidden;
      white-space: nowrap;
      > span {
        width: 90%;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
        display: block;
      }
    }
  }
  ${props =>
    props.error && props.touched
      ? css`
          input[type='file'] {
            + label {
              border: 1px solid ${Color.red};
            }
          }
        `
      : css`
          input[type='file'] {
            + label {
              border: 1px solid ${Color.grey191};
            }
          }
        `}
`;

const IconUpload = styled.div`
  position: absolute;
  top: 34px;
  right: 16px;
`;

export { InputWrapper, IconUpload };
