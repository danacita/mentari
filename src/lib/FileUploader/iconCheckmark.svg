<svg xmlns="http://www.w3.org/2000/svg" width="11" height="8" viewBox="0 0 11 8">
    <path fill="none" fill-rule="evenodd" stroke="green" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M1 3.305L4.348 7 10 1"/>
</svg>
