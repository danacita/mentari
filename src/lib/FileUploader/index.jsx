/* eslint-disable no-nested-ternary */
/* eslint-disable no-return-assign */
/* eslint-disable jsx-a11y/label-has-for */
import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import Color from '../Styles/bases/Color';
import { InputWrapper, IconSuccess } from './Styled';
import Check from './iconCheckmark.svg';

const propTypes = {
  id: PropTypes.string.isRequired,
  accept: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  defaultLabel: PropTypes.array,
  helpBlock: PropTypes.string,
  withRedux: PropTypes.bool,
  prefix: PropTypes.node,
  input: PropTypes.shape({
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.any]),
    onBlur: PropTypes.func,
    onChange: PropTypes.func
  }),
  meta: PropTypes.shape({
    error: PropTypes.string,
    touched: PropTypes.bool
  }),
  uploadDocument: PropTypes.func,
  convertFileToBase64: PropTypes.func
};

const defaultProps = {
  defaultLabel: [],
  helpBlock: null,
  prefix: null,
  input: {},
  meta: {},
  withRedux: true,
  uploadDocument: null,
  convertFileToBase64: null
};

class FileUploader extends Component {
  constructor(props) {
    super(props);
    this.state = {
      fileName: undefined
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleChangeWithoutRedux = this.handleChangeWithoutRedux.bind(this);
  }

  handleChange() {
    // id variable refers to the field ID
    const { id, input, uploadDocument, convertFileToBase64 } = this.props;
    const file = this.fileUpload.files[0];

    // If empty file uploaded, reset the field
    if (!file) {
      input.onChange({});
      return;
    }

    const fileName = file.name;

    // Set state to persist name of file
    this.setState({ fileName });

    // Set field to store uploaded file data
    convertFileToBase64(file).then(base64Data => {
      input.onChange({ file: base64Data, name: fileName });
      uploadDocument(id, base64Data, fileName);
      return Promise.resolve();
    });
  }

  handleChangeWithoutRedux() {
    const file = this.fileUpload.files[0];
    const fileName = file.name;
    // Set state to persist name of file
    this.setState({ fileName });
  }

  render() {
    const {
      id,
      accept,
      label,
      defaultLabel,
      helpBlock,
      prefix,
      withRedux,
      meta: { error, touched }
    } = this.props;
    const { fileName } = this.state;

    return (
      <Fragment>
        {withRedux ? (
          <InputWrapper touched={touched} error={error}>
            <p htmlFor={id}>{label}</p>
            <IconSuccess>
              {fileName === undefined ? (
                defaultLabel && defaultLabel.length !== 0 ? (
                  <img src={Check} alt="Icon" width="16" />
                ) : (
                  <React.Fragment>{prefix}</React.Fragment>
                )
              ) : (
                <img src={Check} alt="Icon" width="16" />
              )}
            </IconSuccess>
            <input
              id={id}
              type="file"
              onChange={this.handleChange}
              ref={ref => (this.fileUpload = ref)}
              accept={accept}
            />
            <label htmlFor={id} error={error}>
              <span>
                {fileName === undefined
                  ? defaultLabel && defaultLabel.length !== 0
                    ? defaultLabel &&
                      defaultLabel.length !== 0 &&
                      defaultLabel[defaultLabel.length - 1].attachment
                        .split('/')[4]
                        .match(`(?<=%2F)(.*)(?=\\?)`)[0]
                    : label
                  : fileName}
              </span>
            </label>
            {touched && error && <span style={{ color: Color.red }}>{error}</span>}
            {helpBlock ? (
              <p className="help-block">
                *&nbsp;
                {helpBlock}
              </p>
            ) : null}
          </InputWrapper>
        ) : (
          <InputWrapper>
            <p htmlFor={id}>{label}</p>
            <IconSuccess withRedux={withRedux}>
              {fileName === undefined ? (
                defaultLabel && defaultLabel.length !== 0 ? (
                  <img src={Check} alt="Icon" width="16" />
                ) : (
                  <React.Fragment>{prefix}</React.Fragment>
                )
              ) : (
                <img src={Check} alt="Icon" width="16" />
              )}
            </IconSuccess>
            <input
              id={id}
              type="file"
              onChange={this.handleChangeWithoutRedux}
              ref={ref => (this.fileUpload = ref)}
              accept={accept}
            />
            <label htmlFor={id}>
              <span>
                {fileName === undefined
                  ? defaultLabel && defaultLabel.length !== 0
                    ? defaultLabel &&
                      defaultLabel.length !== 0 &&
                      defaultLabel[defaultLabel.length - 1].attachment
                        .split('/')[4]
                        .match(`(?<=%2F)(.*)(?=\\?)`)[0]
                    : label
                  : fileName}
              </span>
            </label>
          </InputWrapper>
        )}
      </Fragment>
    );
  }
}

FileUploader.propTypes = propTypes;
FileUploader.defaultProps = defaultProps;

export default FileUploader;
