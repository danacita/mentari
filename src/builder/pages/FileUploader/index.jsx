import React, { Component } from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { atomOneDarkReasonable } from 'react-syntax-highlighter/dist/esm/styles/hljs';
import { ContentWithRightBar } from '../Styled';
import RightbarShared from '../../shared/Rightbar';
import DemoShared from '../../shared/Demo';
import PropsShared from '../../shared/Props';
import CodeShared from '../../shared/Code';
import { Margin } from '../../../lib';

class FileUploaderPage extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const codeString = `import { FileUploader } from '@danacita/mentari';\n<FileUploader/>`;
    return (
      <ContentWithRightBar>
        <h1>File uploader</h1>
        <p>Lorem ipsum dolor, sit amet consectetur adipisicing elit.</p>
        <Margin top={24}>
          <PropsShared>
            <li>
              <span>id</span>
              <span>Apply a id to the control element</span>
            </li>
          </PropsShared>
        </Margin>
        <DemoShared>
          <h4>Demo</h4>
          <p>You can change the props via the component builder on your right</p>
          <Margin top={16}>
            <h4>Demo of file uploader</h4>
          </Margin>
        </DemoShared>
        <h4>Code</h4>
        <Margin top={16}>
          <CodeShared copyCode={codeString}>
            <SyntaxHighlighter language="jsx" style={atomOneDarkReasonable} showLineNumbers>
              {codeString}
            </SyntaxHighlighter>
          </CodeShared>
        </Margin>
        <RightbarShared title="File uploader">
          <Margin bottom={16}>
            <h4>State of File uploader</h4>
          </Margin>
        </RightbarShared>
      </ContentWithRightBar>
    );
  }
}

export default FileUploaderPage;
