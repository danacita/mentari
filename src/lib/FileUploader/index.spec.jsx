import React from 'react';
import renderer from 'react-test-renderer';

import FileUploader from '.';

describe('FileUploader', () => {
  it('renders correctly', () => {
    const input = {
      value: 'Value',
      onBlur: () => {},
      onChange: () => {}
    };

    const meta = {
      error: '',
      touched: false
    };

    const tree = renderer
      .create(
        <FileUploader
          id="fileUploader"
          label="fileUploader"
          input={input}
          meta={meta}
          uploadDocument={() => {}}
          convertFileToBase64={() => {}}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
