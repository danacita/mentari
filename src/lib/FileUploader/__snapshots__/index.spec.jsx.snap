// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploader renders correctly 1`] = `
<div
  className="sc-bdVaJa MXRVf"
>
  <div
    className="sc-bwzfXH eWzXnO"
  />
  <p
    htmlFor="fileUploader"
  >
    fileUploader
  </p>
  <div
    className="sc-htpNat bQadWF"
  />
  <input
    accept=".jpg, .jpeg, .png"
    id="fileUploader"
    onChange={[Function]}
    type="file"
  />
  <label
    error=""
    htmlFor="fileUploader"
  >
    <span>
      fileUploader
    </span>
  </label>
</div>
`;
